internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -299,6 +300,8 @@ func (r *PersistentVolumeClaimReconciler) determineScheduleAndRequeue(
 //     PVCs without ReclaimSpace annotations will be enqueued.
 //   - If the StorageClass has KeyRotation annotation,
 //     PVCs without the KeyRotation annotation will be enqueued.
+//   - If the StorageClass has KeyRotation disable annotation,
+//     PVCs without the KeyRotation disable annotation will be enqueued.
 func (r *PersistentVolumeClaimReconciler) storageClassEventHandler() handler.EventHandler {
```

Comment:
```go
func (r *PersistentVolumeClaimReconciler) storageClassEventHandler() handler.EventHandler {
	return handler.EnqueueRequestsFromMapFunc(
		func(ctx context.Context, obj client.Object) []reconcile.Request {
			ok := false
			obj, ok = obj.(*storagev1.StorageClass)
			if !ok {
				return nil
			}

			// get all PVCs with the same storageclass.
			pvcList := &corev1.PersistentVolumeClaimList{}
			err := r.List(ctx, pvcList, client.MatchingFields{"spec.storageClassName": obj.GetName()})
			if err != nil {
				log.FromContext(ctx).Error(err, "Failed to list PVCs")

				return nil
			}

			// Define the annotations to check
			annotations := []string{rsCronJobScheduleTimeAnnotation, krcJobScheduleTimeAnnotation}

			var requests []reconcile.Request
			for i, _ := range pvcList.Items {
				if shouldEnqueue(obj.GetAnnotations(), pvcList.Items[i].GetAnnotations(), annotations) {
					requests = append(requests, reconcile.Request{
						NamespacedName: types.NamespacedName{
							Name:      pvcList.Items[i].Name,
							Namespace: pvcList.Items[i].Namespace,
						},
					})
				}
			}

			return requests
		},
	)
}

// shouldEnqueue determines whether a PVC should be enqueued based on annotation differences
func shouldEnqueue(scAnnotations, pvcAnnotations map[string]string, keys []string) bool {
	for _, key := range keys {
		if _, scHasAnnotation := scAnnotations[key]; scHasAnnotation {
			if _, pvcHasAnnotation := pvcAnnotations[key]; !pvcHasAnnotation {
				return true
			}
		}
	}
	return false
}

// SetupWithManager sets up the controller with the Manager.
func (r *PersistentVolumeClaimReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions controller.Options) error {
	if err := r.setupIndexers(mgr); err != nil {
		return err
	}

	// Predicate to reconcile on annotation changes
	pred := r.createAnnotationPredicate(rsCronJobScheduleTimeAnnotation, krcJobScheduleTimeAnnotation)
	scPred := r.createAnnotationPredicate(rsCronJobScheduleTimeAnnotation, krcJobScheduleTimeAnnotation)

	return ctrl.NewControllerManagedBy(mgr).
		For(&corev1.PersistentVolumeClaim{}).
		Watches(
			&storagev1.StorageClass{},
			r.storageClassEventHandler(),
			builder.WithPredicates(scPred),
		).
		Owns(&csiaddonsv1alpha1.ReclaimSpaceCronJob{}).
		Owns(&csiaddonsv1alpha1.EncryptionKeyRotationCronJob{}).
		WithEventFilter(pred).
		WithOptions(ctrlOptions).
		Complete(r)
}

// setupIndexers sets up all the necessary field indexers for the reconciler.
func (r *PersistentVolumeClaimReconciler) setupIndexers(mgr ctrl.Manager) error {
	indexes := []struct {
		obj     client.Object
		field   string
		indexFn client.IndexerFunc
	}{
		{
			obj:     &csiaddonsv1alpha1.ReclaimSpaceCronJob{},
			field:   jobOwnerKey,
			indexFn: extractOwnerNameFromPVCObj,
		},
		{
			obj:   &corev1.PersistentVolumeClaim{},
			field: "spec.storageClassName",
			indexFn: func(rawObj client.Object) []string {
				pvc, ok := rawObj.(*corev1.PersistentVolumeClaim)
				if !ok || pvc.Spec.StorageClassName == nil {
					return nil
				}
				return []string{*pvc.Spec.StorageClassName}
			},
		},
		{
			obj:     &csiaddonsv1alpha1.EncryptionKeyRotationCronJob{},
			field:   jobOwnerKey,
			indexFn: r.extractPVCNameFromOwner,
		},
	}

	for _, index := range indexes {
		if err := mgr.GetFieldIndexer().IndexField(context.Background(), index.obj, index.field, index.indexFn); err != nil {
			return err
		}
	}
	return nil
}

// createAnnotationPredicate returns a predicate.Funcs object that triggers reconciliation if any of the provided annotations change.
func (r *PersistentVolumeClaimReconciler) createAnnotationPredicate(annotations ...string) predicate.Funcs {
	return predicate.Funcs{
		UpdateFunc: func(e event.UpdateEvent) bool {
			return r.annotationsChanged(e.ObjectOld.GetAnnotations(), e.ObjectNew.GetAnnotations(), annotations...)
		},
	}
}

// annotationsChanged checks if any of the specified annotations have changed between the old and new objects.
func (r *PersistentVolumeClaimReconciler) annotationsChanged(oldAnnotations, newAnnotations map[string]string, keys ...string) bool {
	for _, key := range keys {
		oldValue, oldExists := oldAnnotations[key]
		newValue, newExists := newAnnotations[key]

		if oldExists != newExists || oldValue != newValue {
			return true
		}
	}
	return false
}
```

can you please refractor the existing code like above (un tested code) and add this new functionality on top of it, this is removing lot of if else blocks

Reply:
lets first do the refactoring in a different PR first and update this PR later on to adopt to the new refactored changes?

Reply:
Makes sense. Let’s do that.